presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java

```diff
@@ -0,0 +1,112 @@
+import static org.apache.parquet.column.ParquetProperties.WriterVersion.PARQUET_1_0;
+import static org.apache.parquet.hadoop.ParquetOutputFormat.COMPRESSION;
+import static org.apache.parquet.hadoop.ParquetOutputFormat.ENABLE_DICTIONARY;
+import static org.apache.parquet.hadoop.ParquetOutputFormat.WRITER_VERSION;
+import static org.apache.parquet.schema.MessageTypeParser.parseMessageType;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+
+public class TestReader
+{
+    @Test
+    public void testTimestampBackedByInt64()
+            throws Exception
+    {
+        MessageType parquetSchema = parseMessageType("message hive_timestamp { optional int64 test (TIMESTAMP_MILLIS); }");
+        ContiguousSet<Long> epochMillisValues = ContiguousSet.create(Range.closedOpen((long) -1_000, (long) 1_000), DiscreteDomain.longs());
+        ImmutableList.Builder<SqlTimestamp> timestamps = new ImmutableList.Builder<>();
+        for (long value : epochMillisValues) {
+            timestamps.add(SqlTimestamp.fromMillis(3, value));
+        }
+
+        List<ObjectInspector> objectInspectors = singletonList(javaLongObjectInspector);
+        List<String> columnNames = ImmutableList.of("test");
+
+        ConnectorSession session = getHiveSession(new HiveConfig()
+                .setHiveStorageFormat(HiveStorageFormat.PARQUET)
+                .setUseParquetColumnNames(true));
+
+        try (ParquetTester.TempFile tempFile = new ParquetTester.TempFile("test", "parquet")) {
+            JobConf jobConf = new JobConf();
+            jobConf.setEnum(COMPRESSION, CompressionCodecName.SNAPPY);
+            jobConf.setBoolean(ENABLE_DICTIONARY, true);
+            jobConf.setEnum(WRITER_VERSION, PARQUET_1_0);
+
+            ParquetTester.writeParquetColumn(
+                    jobConf,
+                    tempFile.getFile(),
+                    CompressionCodecName.SNAPPY,
+                    ParquetTester.createTableProperties(columnNames, objectInspectors),
+                    getStandardStructObjectInspector(columnNames, objectInspectors),
+                    new Iterator<?>[] {epochMillisValues.iterator()},
+                    Optional.of(parquetSchema),
+                    false);
+
+            Iterator<SqlTimestamp> expectedValues = timestamps.build().iterator();
+            try (ConnectorPageSource pageSource = FileFormat.PRESTO_PARQUET.createFileFormatReader(session, HDFS_ENVIRONMENT, tempFile.getFile(), columnNames, ImmutableList.of(TIMESTAMP_MILLIS))) {
+                // skip a page to exercise the decoder's skip() logic
+                Page firstPage = pageSource.getNextPage();
+                for (int i = 0; i < firstPage.getPositionCount(); i++) {
+                    expectedValues.next();
+                }
+
+                while (!pageSource.isFinished()) {
+                    Page page = pageSource.getNextPage();
+                    if (page == null) {
```

Comment:
We should document what null return value means. "This method is allowed to return null" is not sufficient.

Reply:
Agreed, but unrelated to this PR